src/Dhall/Map.hs

```diff
@@ -0,0 +1,258 @@
+{-# LANGUAGE CPP                #-}
+{-# LANGUAGE DeriveDataTypeable #-}
+{-# LANGUAGE DeriveTraversable  #-}
+{-# LANGUAGE RecordWildCards    #-}
+
+-- | `Map` type used to represent records and unions
+
+module Dhall.Map
+    ( -- * Type
+      Map
+
+      -- * Construction
+    , singleton
+    , fromList
+
+      -- * Sorting
+    , sort
+    , isSorted
+
+      -- * Insertion
+    , insert
+    , insertWith
+
+      -- * Deletion/Update
+    , delete
+
+      -- * Query
+    , lookup
+    , member
+
+      -- * Combine
+    , union
+    , unionWith
+    , intersection
+    , intersectionWith
+
+      -- * Traversals
+    , mapWithKey
+    , traverseWithKey
+    , foldMapWithKey
+
+      -- * Conversions
+    , toList
+    , toMap
+    , keys
+    ) where
+
+import Control.Applicative ((<|>))
+import Data.Data (Data)
+import Data.Semigroup
+import Prelude hiding (lookup)
+
+import qualified Data.Map
+import qualified Data.Set
+
+{-| A `Map` that remembers the original ordering of keys
+
+    This is primarily used so that formatting preserves field order
+
+    This is done primarily to avoid a dependency on @insert-ordered-containers@
+    and also to improve performance
+-}
+data Map k v = Map (Data.Map.Map k v) [k]
```

Comment:
`Seq` instead of a list improves the asymptotic of `union`.

Reply:
I'm going to add a benchmark first to double-check this because sometimes the better per-element constant factors of lists outweigh the asymptotic penalty